### DISCUSSION TITLE
Enum mapping from String to specific enum type, but generalized.

### DISCUSSION THREAD

**Discussion by DmitrijG934** [mapstruct/mapstruct#3777] [Q&A] [answered]
Hello, folks. Just wondering: is there any way to configure MapStruct using some kind of global configuration class or properties, or whatever to be able to tell the framework how to map string values to enums (ignoring enum types)?

I'm aware about possibility to use SPI and define my own EnumTransformatingStrategy, but I think for my project it will be too complicated. 
Let me explain the problem in detail:

1) I have a Spring Boot microservice, it uses generated via OpenAPI generator servers and models, MapStruct for mapping between API models and BLL models.
2)  I have the following enums (generated via OpenAPI generator):

```
enum Vat {
  _1("-1"), _2("-2");
  
  private String value;
}

enum OrderType {
  DELIVERY("OrderDelivery"), PACKAGE("OrderPackage");

  private String value;
}
```

3) The problem is that MapStruct uses `Enum.valueOf( Vat.class, order.getVat() )`, so I wrote my own `default` method in order to resolve the issue with incorrect mapping:

The general idea here is to use enum values for mapping, not enum constant name.

Example with String source type;

```
default Vat mapStringToVatEnum(String source) {
        return ofNullable(source)
                .map(value -> {
                    try {
                        return Vat.fromValue(value);
                    } catch (IllegalArgumentException e) {
                        return Vat.valueOf(value);
                    }
                })
                .orElse(null);
    }
```

Example with another enum:

```
    default package1.VatEnum mapVatEnumToEnum(package2.VatEnum source) {
        return ofNullable(source)
                .map(value -> {
                    try {
                        return package1.VatEnum.fromValue(value.getValue());
                    } catch (IllegalArgumentException e) {
                        return package1.VatEnum.valueOf(value.name());
                    }
                })
                .orElse(null);
    }

```

Can someone to assist me with more flexible and elegant solution? Thanks in advance.

**Comment by filiphr** [comment, marked as answer]
@DmitrijG934 there have been some suggestions in the issues to have an annotation that we can use in the `enum` that would tell MapStruct to actually use the `fromValue` method in the enum.

What you can do now is to have a custom method and register that with MapStruct through `Mapper#uses` or `MapperConfig#uses`

e.g.

```java
public class MyEnumMapper {

    public static Vat mapStringToVatEnum(String source) {
        return ofNullable(source)
                .map(value -> {
                    try {
                        return Vat.fromValue(value);
                    } catch (IllegalArgumentException e) {
                        return Vat.valueOf(value);
                    }
                })
                .orElse(null);
    }

}
```

and then in your own mapper you can do 

```java
@Mapper(uses = MyEnumMapper.class)
public class MyMapper {

    //....

}
```

As for the enum to enum, I don't fully understand why you would need to do the `getValue` and `fromValue`. With that one you can easily just use a regular [enum to enum mapping](https://mapstruct.org/documentation/stable/reference/html/#_mapping_enum_to_enum_types)

**Reply by DmitrijG934** [reply]
Thanks for the answer, @filiphr. Looks useful for me.
